### PR TITLE
Trade with some or all of your offchain funds after settling

### PR DESCRIPTION
**Pull request by luckysori**
Fix https://github.com/get10101/10101/issues/1781.
Fix[^1] https://github.com/get10101/10101/issues/1802.
Fix https://github.com/get10101/10101/issues/1798 (missing rollover, which I will reintroduce separately, and collab revert, which will be done in https://github.com/get10101/10101/pull/1823).

Some TODOs for now or later:

- [x] Add e2e test.
- [x] Fix offchain wallet balance. I think we might need to model this explicitly and update after every update to the DLC channel based on some stored metadata. Seems insufficient to rely on the `rust-dlc` DLC `Channel`, because the margin and collateral reserve split is only modeled in the payout curve.
- [ ] ~Tackle https://github.com/get10101/10101/issues/1791 since it should be straightforward now.~ (Later.)

[^1]: In the sense that I've added code to prevent running into the error when handling a message, and instead we fail early e.g. https://github.com/get10101/10101/pull/1811/files#diff-96878005a821426ca9541b70172d80d383ebaed2c819bb4b7216229c7b56492aR29-R45.

### CODE REVIEW

**Comment by luckysori** [comment]
Proof that I didn't break the `payout_curve` example:

image

**Comment by bonomat** [comment]
Uff.. this is hard to review :/ 

I think I found one problem when playing around with it: 
I think the `trader_reserve` is not calculated correctly when opening a new position: 

Here is what I did: 

1. BTCUSD: 40,000
2. Quantity: 10
3. Leverage: 2
4. Needed margin: 12,500
5. Opening fee: 75

--> Funded the channel with 12,575

then I closed the channel and ended up with `12,425` sats in my channel which seems to be correct, because I've paid another 75 sats closing fee, i.e. `12,575`-`75 opening fee` - `75 closing fee` = `12,425`

Then I opened a new position


1. BTCUSD: 40,000
2. Quantity: 8
3. Leverage: 2
4. Needed margin: 10,000
5. Opening fee: 60

--> Needed 10,060 

Then I closed the channel again and ended up with `9,940` (10,000-60 opening fee - 60 closing fee) but I've lost the remaining money I had before, i.e. `12,425-9,940=2,485`sats are missing

**Comment by luckysori** [comment]
> Uff.. this is hard to review :/

I'm sorry, but I don't think I could have done much better with the task at hand. The second to last commit should be okay, although I had to leave some `unimplemented`s in, which I do not like. Maybe I can clean up the last commit a little bit, but I don't think it will make a big difference.

> I think I found one problem when playing around with it: I think the `trader_reserve` is not calculated correctly when opening a new position:
> 
> Here is what I did:
> 
>     1. BTCUSD: 40,000
> 
>     2. Quantity: 10
> 
>     3. Leverage: 2
> 
>     4. Needed margin: 12,500
> 
>     5. Opening fee: 75
> 
> 
> --> Funded the channel with 12,575
> 
> then I closed the channel and ended up with `12,425` sats in my channel which seems to be correct, because I've paid another 75 sats closing fee, i.e. `12,575`-`75 opening fee` - `75 closing fee` = `12,425`
> 
> Then I opened a new position
> 
>     1. BTCUSD: 40,000
> 
>     2. Quantity: 8
> 
>     3. Leverage: 2
> 
>     4. Needed margin: 10,000
> 
>     5. Opening fee: 60
> 
> 
> --> Needed 10,060
> 
> Then I closed the channel again and ended up with `9,940` (10,000-60 opening fee - 60 closing fee) but I've lost the remaining money I had before, i.e. `12,425-9,940=2,485`sats are missing

Thanks for testing the pull request! I will follow this path when I add the e2e test and make sure that this is fixed.

File: mobile/native/src/trade/order/handler.rs [hunk 1]
```diff
@@ -23,6 +26,24 @@ use uuid::Uuid;
 const ORDER_OUTDATED_AFTER: Duration = Duration::minutes(5);
 
 pub async fn submit_order(order: Order) -> Result<Uuid> {
+    // If we have an open position, We should not allow any further trading until the current DLC
+    // channel is confirmed on-chain. Otherwise we can run into pesky DLC protocol failures.
```
Comment (bonomat) [inline mobile/native/src/trade/order/handler.rs:30]: TIL: pesky 😅

**Review by bonomat — COMMENTED** [review]
Looks all correct to me. I'm not sure where the flaw is which I documented earlier. I just added a few comments above.

File: coordinator/src/node.rs [hunk 2]
```diff
@@ -203,98 +203,205 @@ impl Node {
...
-            }
-            TradeAction::Resize(channel_id) => {
-                tracing::debug!(trader_id, order_id, "Resizing existing position");
-                ensure!(
-                    self.settings.read().await.allow_opening_positions,
-                    "Resizing positions is disabled"
-                );
+        tracing::info!(
+            %peer_id,
+            order_id = %trade_params.filled_with.order_id,
```
[the remark this answers is not in the document]
Comment (luckysori) [reply coordinator/src/node.rs:227]: Yep, the `order_id` is a `Uuid` which comes with a nice `Display` implementation.

File: crates/ln-dlc-node/src/node/dlc_channel.rs [hunk 3]
```diff
@@ -290,6 +290,25 @@ impl<S: TenTenOneStorage + 'static, N: LnDlcStorage + Sync + Send + 'static> Nod
         .map_err(|e| anyhow!("{e:#}"))?
     }
 
+    /// Accept an update to the DLC channel. This can only succeed if we previously received an DLC
+    /// channel update offer from the the counterparty.
+    pub fn accept_dlc_channel_update(&self, channel_id: &DlcChannelId) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (luckysori) [reply crates/ln-dlc-node/src/node/dlc_channel.rs:306]: Great catch! I had inlined this in `mobile/native/src/ln_dlc/node.rs`. Fixing it.

File: crates/payout_curve/src/lib.rs [hunk 4]
```diff
@@ -30,6 +32,93 @@ pub struct PayoutPoint {
     pub extra_precision: u16,
 }
 
+#[derive(Clone, Copy)]
+pub struct PartyParams {
+    /// How many coins the party is wagering.
+    margin: Amount,
```
[the remark this answers is not in the document]
Comment (luckysori) [reply crates/payout_curve/src/lib.rs:38]: I admit that the important thing is that the arguments in the constructor are strongly typed. I only kept it because I liked not needing to document the type haha.

I can use `u64` instead, it might be more correct.

File: coordinator/src/node.rs [hunk 5]
```diff
@@ -202,98 +203,205 @@ impl Node {
...
+                        "Trader cannot trade with more than their total collateral in \
+                         the DLC channel: margin ({}) > collateral ({})",
+                        margin_trader, updated_collateral
+                    )
+                })?;
+
+            (updated_collateral, reserve)
+        };
+
+        // TODO: We need to build the `ContractDescriptor` based on the total collateral in the DLC
```
[the remark this answers is not in the document]
Comment (luckysori) [reply coordinator/src/node.rs:394]: Good catch, this was part of an earlier version!

File: coordinator/src/node.rs [hunk 6]
```diff
@@ -202,98 +203,205 @@ impl Node {
...
+        let order_matching_fee = order_matching_fee_taker(
             trade_params.quantity,
             trade_params.average_execution_price(),
         )
         .to_sat();
-        let initial_price = trade_params.filled_with.average_execution_price();
 
         let coordinator_direction = trade_params.direction.opposite();
 
+        let (updated_coordinator_collateral, coordinator_collateral_reserve) = {
```
[the remark this answers is not in the document]
Comment (luckysori) [reply coordinator/src/node.rs:360]: This is a very fair criticism.

How do you feel about using the word "collateral" for the money in the _channel_ and "margin" for the money being traded (thus, less than or equal to the collateral). Originally I thought this was a good idea, but I don't think it works because those two words probably mean the same thing?

[the remark this answers is not in the document]
Comment (luckysori) [reply coordinator/src/node.rs:360]: Maybe what I mentioned in my previous comment is orthogonal.

In any case, I've simplified this so that there is no need for this `updated_collateral` concept.

File: crates/tests-e2e/src/coordinator.rs [hunk 7]
```diff
@@ -23,18 +23,19 @@ pub struct Coordinator {
 }
 
 #[derive(Deserialize)]
-pub struct DlcChannels {
+pub struct SubChannels {
```
Comment (bonomat) [inline crates/tests-e2e/src/coordinator.rs:26]: What am I missing here? Why did you change this to `SubChannels` from `DlcChannels`? 
Might have been a rebase error.

**Review by bonomat — APPROVED** [review]
Looks good. Great work.

**Comment by bonomat** [comment]
@luckysori : I took over and rebase (a couple of times 🙈). I hope I did not mess things up.

File: crates/tests-e2e/src/coordinator.rs (hunk 7 above)
Comment (luckysori) [reply crates/tests-e2e/src/coordinator.rs:26]: rebase error indeed!

**Comment by luckysori** [comment]
> @luckysori : I took over and rebase (a couple of times 🙈). I hope I did not mess things up.

Thanks for taking over and for the fix!

If the GitHub intermediate diffs are to be trusted, it looks good!
